### ISSUE TITLE
New buzzwords Kubernetes for Edge Computing: The Microsoft Azure Approach

### ISSUE THREAD

**Issue by sudocurse** [sudocurse/reading-list#12894]
Kubernetes for Edge Computing: The Microsoft Azure Approach<br>
https://ift.tt/2IdL3sZ<br>
<img width="640" height="412" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/05/5243a3ab-mist-1659115_640.jpg" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2018/05/5243a3ab-mist-1659115_640.jpg">
<p>Edge devices are designed to live at the edge of the network, often on a slow or occasional connection, and increasingly they&rsquo;re used for much more than monitoring. Microsoft has created a managed service, called <a href="https://docs.microsoft.com/en-us/azure/iot-edge/how-iot-edge-works" class="ext-link" rel="external nofollow">Azure IoT Edge</a>, to put analytics and intelligence on those devices using containers. Think of it as a way to move cloud workloads like analytics, machine learning, SQL Server databases and even serverless functions onto devices that are centrally deployed and managed, even if they&rsquo;re only occasionally connected.</p>
<p>As announced at the recent Build conference, those workloads now include Microsoft&rsquo;s Cognitive Services; the first one available is <a href="https://www.customvision.ai/" class="ext-link" rel="external nofollow">Custom Vision</a>, which uses the same ResNet-50 deep learning network that Microsoft Bing uses for image recognition, but with the final layer replaced by one trained against your specific images to do image classification and object recognition (finding and labeling the location of known objects in the image with a bounding box).</p>
<p>For object recognition, you can use the Custom Vision portal to find all the objects that have been recognized and tag them specifically (including adjusting the bounding boxes to delineate objects more precisely to help with the training). Upload as few as 50 labeled images per class as the training set and in less than a minute you have a trained model that you can deploy as a module to an Azure IoT Edge device.</p>
<p>Keep feeding the results of the image recognition back in (or add more training images) and you get a better model that you can update devices with.</p>
<p>What you get in the export is a &lsquo;compact&rsquo; version of the model that runs on more constrained mobile devices and you need to select a compact domain to train the model in before exporting (currently the options are general, landmarks and retail, which isn&rsquo;t quite as many domain options as if you call the model as an API). The tradeoff is that it may not be as accurate as the same data model running in the cloud where there are more resources, but it can run right where you need the recognition done, even if there&rsquo;s no network connection.</p>
<p>The IoT Edge export from the Custom Vision service is a DockerFile, a TensorFlow model and service code that can run in a Windows or Linux container (on X64 or ARM). The Azure IoT Edge runtime is <a href="https://github.com/azure/iot-edge" class="ext-link" rel="external nofollow">open source</a>.</p>
<div id="attachment_4766009" style="width: 710px;" class="wp-caption aligncenter">
<a href="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/05/601336d0-kubernetes-edge.png" class="ext-image" rel="external nofollow"><img class="size-full wp-image-4766009" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/05/601336d0-kubernetes-edge.png" width="700" height="580" data-id="4766009"></a>
<p class="wp-caption-text">Courtesy of Microsoft.</p>
</div>
<p>If you want to create your own data models instead of using Cognitive Services, you can <a href="https://docs.microsoft.com/en-us/azure/iot-edge/tutorial-deploy-machine-learning" class="ext-link" rel="external nofollow">package those</a> as Docker containers to deploy to IoT Edge in the same way. To simplify packaging up models built with Azure Machine Learning, <a href="https://github.com/Azure/ai-toolkit-iot-edge" class="ext-link" rel="external nofollow">use the AI Toolkit for Azure IoT Edge</a>.</p>
<p>And now there are more ways to deploy those containers onto edge devices managed by IoT Edge than just through the Azure portal.</p>
<p>To make it easier to manage IoT Edge deployments from inside a Kubernetes environment like Azure Kubernetes Service (the new, less confusing name for the fully managed Kubernetes container orchestration service on Azure), using familiar Kubernetes concepts like manifests, Microsoft has just published an open source <a href="https://github.com/azure/iot-edge-virtual-kubelet-provider" class="ext-link" rel="external nofollow">IoT Edge Virtual Kubelet</a> provider on GitHub that you can install in your Kubernetes cluster using Helm.</p>
<blockquote>
<p>&ldquo;This gives you a consistent interface where you can push an entire Kubernetes workload across cloud and edge,&rdquo; &mdash; Sam George.</p>
</blockquote>
<p>As <a href="https://www.linkedin.com/in/samjgeorge/" class="ext-link" rel="external nofollow">Sam George</a>, director&nbsp;of Azure IoT engineering, explained to The New Stack at Build, &ldquo;In a Kubernetes cluster, Kubernetes nodes are typically bound to a VM; if I push containers to the cluster, they get placed inside a VM. A <a href="https://github.com/virtual-kubelet/virtual-kubelet/blob/master/README.md" class="ext-link" rel="external nofollow">Virtual Kubelet</a> works like a proxy. It can interpret the Kubernetes commands and do something else with [the container] beside placing it in a VM.&rdquo;</p>
<p>So far, Virtual Kubelets have been about bursting out from a Kubernetes cluster to a cloud Kubernetes service like Azure Container Instances or <a href="https://github.com/hyperhq" class="ext-link" rel="external nofollow">Hyper</a> to get an advantage like scale or speed or per-second billing without having to worry about managing agent nodes. The ACI Connector is a Virtual Kubelet. Here, though, the IoT Edge Virtual Kubelet provider is placing the containers into edge devices that then look as if they&rsquo;re part of the cluster.</p>
<p>Host the provider in a Kubernetes cluster and it shows up as just another node, George explained. &ldquo;When you push containers to the IoT Edge provider we interpret that push as an IoT Edge deployment; we go through our Azure IoT Hub service and push that container out to an edge device. The Kubernetes master thinks it&rsquo;s just got a big Kubernetes cluster running in the cloud but in fact, some of the containers are now running on edge devices.&rdquo;</p>
<p>The provider is still in development and not all of the integrations are finished, but it promises a lot of flexibility. The big advantage of IoT Edge is bringing cloud services consistently to edge devices, and using Kubernetes makes that apply to more than the Azure services that are integrated with IoT Edge. You can use it to bring data models to edge devices, but if there&rsquo;s a new version of, say, the compression algorithm you use, you can manage to deploy that to the cloud and edge systems in the same way.</p>
<p>The IoT Edge provider can push to multiple edge devices, using tags and device selector query to target the deployment to the correct edge devices. IoT devices with the same software configuration will often be connected to IoT hubs in different regions for performance or resiliency; creating a virtual Kubernetes node for each of those IoT hubs means the same kubectl command can apply deployment manifests to devices with the same configuration even though they&rsquo;re attached to different hubs.</p>
<p>It doesn&rsquo;t matter if those IoT devices are on slow connections, as long as they&rsquo;re connected, George said. &ldquo;The part that doesn&rsquo;t work yet is when there are times when the edge device is disconnected. The Virtual Kubelet works great where there is connectivity to the edge at all time because it&rsquo;s reporting back two-way communication with the health of each container.&rdquo;</p>
<p class="attribution">Microsoft is a sponsor of The New Stack.</p>
<p class="attribution">Feature image <a href="https://pixabay.com/en/mist-misty-fog-foggy-morning-lake-1659115/" class="ext-link" rel="external nofollow">via</a> Pixabay.</p>
<p>The post <a rel="nofollow" href="https://thenewstack.io/kubernetes-for-edge-computing-the-microsoft-azure-approach/">Kubernetes for Edge Computing: The Microsoft Azure Approach</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br>
<br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
May 14, 2018 at 11:57AM
